### PR TITLE
[flang][openacc] Check trip count invariance with other IVs

### PR DESCRIPTION
**Pull request by clementval**
2.9.1 The trip count for all loops associated with the collapse clause must be computable and invariant in all the loops.

This patch checks that iteration range of loops part of a collapse nest does not depend on outer loops induction variables.

The check is also applied to combined construct with a loop.

### CODE REVIEW

File: flang/lib/Semantics/resolve-directives.cpp [hunk 1]
```diff
@@ -1250,10 +1256,41 @@ void AccAttributeVisitor::CheckAssociatedLoopIndex(
...
+
+  Symbol::Flag flag;
+  llvm::SmallVector<Symbol *> ivs;
+  using Bounds = parser::LoopControl::Bounds;
+  for (const parser::DoConstruct *loop{&outerDoConstruct}; loop && level > 0;) {
     // Go through all nested loops to ensure index variable exists.
-    GetLoopIndex(*loop);
+    if (const parser::Name * ivName{GetLoopIndex(*loop)}) {
+      if (auto *symbol{ResolveAcc(*ivName, flag, currScope())}) {
+        if (auto &control = loop->GetLoopControl()) {
```
Comment (vzakhari) [inline flang/lib/Semantics/resolve-directives.cpp:1276]: nit: I see a mix of var initialization styles (via `{}` and `=`).  You may consider sticking to one :)

**Review by vzakhari — APPROVED** [review]
LGTM. Thank you!

File: flang/lib/Semantics/resolve-directives.cpp (hunk 1 above)
Comment (clementval) [reply flang/lib/Semantics/resolve-directives.cpp:1276]: Thanks for catching that! I'll update

**Review by razvanlupusoru — APPROVED** [review]
Nice!

File: flang/lib/Semantics/resolve-directives.cpp (hunk 1 above)
Comment (clementval) [inline flang/lib/Semantics/resolve-directives.cpp:1276]:
```suggestion
        if (auto &control{loop->GetLoopControl()}) {
```

File: flang/lib/Semantics/resolve-directives.cpp [hunk 2]
```diff
@@ -1250,10 +1256,41 @@ void AccAttributeVisitor::CheckAssociatedLoopIndex(
...
+  llvm::SmallVector<Symbol *> ivs;
+  using Bounds = parser::LoopControl::Bounds;
+  for (const parser::DoConstruct *loop{&outerDoConstruct}; loop && level > 0;) {
     // Go through all nested loops to ensure index variable exists.
-    GetLoopIndex(*loop);
+    if (const parser::Name * ivName{GetLoopIndex(*loop)}) {
+      if (auto *symbol{ResolveAcc(*ivName, flag, currScope())}) {
+        if (auto &control{loop->GetLoopControl()}) {
+          if (const Bounds * b{std::get_if<Bounds>(&control->u)}) {
+            if (auto lowerExpr = semantics::AnalyzeExpr(context_, b->lower)) {
```
Comment (clementval) [inline flang/lib/Semantics/resolve-directives.cpp:1278]:
```suggestion
            if (auto lowerExpr{semantics::AnalyzeExpr(context_, b->lower)}) {
```

File: flang/lib/Semantics/resolve-directives.cpp [hunk 3]
```diff
@@ -1250,10 +1256,41 @@ void AccAttributeVisitor::CheckAssociatedLoopIndex(
...
+    if (const parser::Name * ivName{GetLoopIndex(*loop)}) {
+      if (auto *symbol{ResolveAcc(*ivName, flag, currScope())}) {
+        if (auto &control{loop->GetLoopControl()}) {
+          if (const Bounds * b{std::get_if<Bounds>(&control->u)}) {
+            if (auto lowerExpr{semantics::AnalyzeExpr(context_, b->lower)}) {
+              semantics::UnorderedSymbolSet lowerSyms =
+                  evaluate::CollectSymbols(*lowerExpr);
+              checkExprHasSymbols(ivs, lowerSyms);
+            }
+            if (auto upperExpr = semantics::AnalyzeExpr(context_, b->upper)) {
```
Comment (clementval) [inline flang/lib/Semantics/resolve-directives.cpp:1283]:
```suggestion
            if (auto upperExpr{semantics::AnalyzeExpr(context_, b->upper)}) {
```

**Pull request merged**